pkg/assets/fetch_asset.go

```diff
@@ -0,0 +1,227 @@
+package assets
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/go-openapi/swag"
+	"github.com/labstack/gommon/log"
+	"github.com/massalabs/station-massa-wallet/api/server/models"
+	"github.com/pkg/errors"
+)
+
+const assetsJSONPath = "assets.json"
+
+type assetData struct {
+	Assets []struct {
+		ContractAddress string `json:"contractAddress"`
+		Name            string `json:"name"`
+		Symbol          string `json:"symbol"`
+		Decimals        int64  `json:"decimals"`
+	} `json:"assets"`
+}
+
+
+// AssetsStore encapsulates the contract assets and mutex.
+type AssetsStore struct {
+	ContractAssets      map[string]models.AssetInfo
+	contractAssetsMutex sync.Mutex
+}
+
+// RetrieveAssetsInfo retrieves asset information for a given list of contract addresses concurrently.
+func (s *AssetsStore) RetrieveAssetsInfo(contractAddresses []string) ([]models.AssetInfo, error) {
+	var wg sync.WaitGroup
+	assetInfoCh := make(chan *models.AssetInfo, len(contractAddresses))
+	errCh := make(chan error, len(contractAddresses))
+
+	for _, address := range contractAddresses {
+		wg.Add(1)
+		go func(address string) {
+			defer wg.Done()
+
+			assetInfo, err := s.RetrieveAssetInfo(address)
+			if err != nil {
+				errCh <- errors.Wrapf(err, "failed to retrieve asset information for contract address %s", address)
+				return
+			}
+
+			assetInfoCh <- assetInfo
+		}(address)
+	}
+
+	wg.Wait()
+	close(assetInfoCh)
+	close(errCh)
+
+	// Collect asset information and errors
+	var assetInfos []models.AssetInfo
+	var allErrors []error
+
+	for assetInfo := range assetInfoCh {
+		assetInfos = append(assetInfos, *assetInfo)
+	}
+
+	for err := range errCh {
+		allErrors = append(allErrors, err)
+	}
+
+	// If there were errors, return them along with the retrieved asset information
+	if len(allErrors) > 0 {
+		return assetInfos, fmt.Errorf("failed to retrieve some assets: %w", gatherErrors(allErrors))
+	}
+
+	return assetInfos, nil
+}
+
+// RetrieveAssetInfo retrieves the asset information for a given contract address.
+func (s *AssetsStore) RetrieveAssetInfo(contractAddress string) (*models.AssetInfo, error) {
+	// Check if the asset information exists in the JSON
+	assetInfo, err := s.AssetInfo(contractAddress)
+	if err == nil {
+		// Asset information found in the JSON
+		return assetInfo, nil
+	} else if errors.Is(err, ErrAssetInfoNotFound) {
+		// If asset information is not found in the JSON, call the smart contract to retrieve it
+		assetInfoFromSC, err := s.getAssetInfoFromSC(contractAddress)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to retrieve asset information from smart contract")
+		}
+
+		// Update the ContractAssets map with the new asset information
+		s.contractAssetsMutex.Lock()
+		defer s.contractAssetsMutex.Unlock()
+
+		s.ContractAssets[contractAddress] = *assetInfoFromSC
+
+		// Persist the updated ContractAssets map back to the JSON file
+		if err := s.persistAssets(); err != nil {
```

Comment:
why we persist the asset here? its confusing.. its mean we have to call RetrieveAssetInfo to persist an asset .. it works but its weird logic